crates/storage-api/src/vqueue_table/metadata.rs

```diff
@@ -0,0 +1,328 @@
+    #[bilrost(4)]
+    pub(crate) last_completion_at: Option<UniqueTimestamp>,
+    #[bilrost(5)]
+    pub(crate) num_running: u32,
+}
+
+impl VQueueStatistics {
+    fn update_avg_queue_duration(&mut self, latency_ms: u64) {
+        let new_avg: u64 = if self.avg_queue_duration_ms == 0 {
+            latency_ms
+        } else {
+            // exponential moving average
+            ((self.avg_queue_duration_ms as f64 * 0.95) + (latency_ms as f64 * 0.05)).ceil() as u64
+        };
+        self.avg_queue_duration_ms = u32::try_from(new_avg).unwrap_or(u32::MAX);
+    }
+
+    fn increment_running(&mut self) {
+        self.num_running += 1;
+    }
+
+    fn decrement_running(&mut self) {
+        self.num_running -= 1;
+    }
+}
+
+#[derive(Debug, Default, Clone, bilrost::Message)]
+pub struct VQueueMeta {
+    /// if true, the vqueue is paused, we don't pop entries from it until it's resumed.
+    #[bilrost(1)]
+    is_paused: bool,
+    /// Total number of entries (ready + paused + running + suspended + scheduled), but it doesn't
+    /// include completed or failed entries. This is the length that is used to reject new invocations
+    /// being added to the vqueue. The capacity configuration will limit this value.
+    /// in-flight can be calculated by length - num_pending. = suspended + paused + running.
+    #[bilrost(2)]
+    pub(crate) stats: VQueueStatistics,
+    #[bilrost(3)]
+    pub(crate) length: u32,
+    /// Number of concurrency tokens being used
+    #[bilrost(4)]
+    pub(crate) num_tokens_used: u32,
+    /// The number of entries waiting to be dequeued. The vector index implies the priority
+    #[bilrost(5)]
+    pub(crate) num_waiting: smallvec::SmallVec<[u32; EffectivePriority::NUM_PRIORITIES]>,
+}
+
+impl VQueueMeta {
+    pub fn tokens_used(&self) -> u32 {
+        self.num_tokens_used
+    }
+
+    pub fn len(&self) -> u32 {
+        self.length
+    }
+
+    pub fn is_empty(&self) -> bool {
+        self.length == 0
+    }
+
+    pub fn total_waiting(&self) -> u32 {
+        self.num_waiting.iter().sum()
+    }
+
+    pub fn status(&self) -> VQueueStatus {
+        if self.is_empty() {
+            VQueueStatus::Empty
+        } else if self.is_paused() {
+            VQueueStatus::Paused
+        } else {
+            VQueueStatus::Active
+        }
+    }
+
+    pub fn num_waiting(&self, priority: EffectivePriority) -> u32 {
+        self.num_waiting
+            .get(priority as usize)
+            .copied()
+            .unwrap_or_default()
+    }
+
+    pub fn num_running(&self) -> u32 {
+        self.stats.num_running
+    }
+
+    pub fn stats(&self) -> &VQueueStatistics {
+        &self.stats
+    }
+
+    pub fn last_enqueued_ts(&self) -> Option<UniqueTimestamp> {
+        self.stats.last_enqueued_at
+    }
+
+    pub fn is_paused(&self) -> bool {
+        self.is_paused
+    }
+
+    fn add_to_waiting(&mut self, priority: EffectivePriority) {
+        self.num_waiting.resize(priority as usize + 1, 0);
```

Comment:
Are we possibly losing values of lower priorities when resizing to a higher priority and then back to a lower priority? E.g. would this pass:

```
self.add_to_waiting(EffectivePriority::UserDefault);
self.add_to_waiting(EffectivePriority::UserHigh);
self.add_to_waiting(EffectivePriority::UserDefault);
assert_eq!(self.num_waiting(EffectivePriority::UserDefault), 2);
```